### ISSUE TITLE
[no-unnecessary-type-assertion] false positive when asserting the result of generic function invocation

### ISSUE THREAD

**Issue by vkrol** [typescript-eslint/typescript-eslint#528]
**Repro**
https://github.com/vkrol/typescript-eslint-no-unnecessary-type-assertion-generic-function

```js
module.exports = {
	parser: "@typescript-eslint/parser",
	parserOptions: {
		project: "tsconfig.json",
	},
	plugins: ["@typescript-eslint"],
	rules: {
		"@typescript-eslint/no-unnecessary-type-assertion": "error",
	}
};
```

```TS
declare function foo<T extends Element = Element>(): T;
const a = (foo() as HTMLInputElement).value;
```
**Expected Result**
No errors.
**Actual Result**
Error:
```
D:\Projects\typescript-eslint-no-unnecessary-type-assertion-generic-function\index.ts
  2:12  error  This assertion is unnecessary since it does not change the type of the expression  @typescript-eslint/no-unnecessary-type-assertion

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```
**Additional Info**
N/A.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.9.0` |
| `@typescript-eslint/parser`        | `1.9.0` |
| `TypeScript`                       | `3.4.5` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `10.15.3` |
| `npm`                              | `6.4.1` |

**Comment by bradzacher** [comment]
Looking into this, it looks like a bug (or a feature?) in typescript.

It is inferring that the function call uses `HTMLInputElement` as a generic:
![image](https://user-images.githubusercontent.com/7462525/57870777-a9edc600-77bc-11e9-855f-a4f546106834.png)

The checker will therefore always report to use that the function's return type as matching the type annotation.

I'm going to close this, as there's nothing we can do about this from our side without adding a large amount of extra code to check if the value being cast comes from a function and that function has a generic etc.

As a workaround you can either delete the generic (I doubt you can, it's probably a library function, right?):
```ts
declare function foo(): Element;
const a = (foo() as HTMLInputElement).value;
```
Or use the generic.
```ts
declare function foo<T extends Element = Element>(): T;
const a = foo<HTMLInputElement>().value;
```

If you feel strongly about it, I would suggest raising an issue in the typescript repo.

**Comment by vkrol** [comment]
Thank you for the investigation.

Yes, it is the library function and I do no want to use the generic parameter because I think that it is the anti-pattern:
>getMeAT(): T: If a type parameter does not appear in the types of any parameters, you don't really have a generic function, you just have a disguised type assertion. Prefer to use a real type assertion, e.g. getMeAT() as number. Example where a type parameter is acceptable: function id(value: T): T;. Example where it is not acceptable: function parseJson(json: string): T;. Exception: new Map<string, number>() is OK.

https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes

I need to think a little more about it.
